### PR TITLE
Display fit results in grouped workspaces and update model in custom setup

### LINKED ISSUE #27780: Engineering Fitting Tab - Running Fits
### Story:
As a user, once I have selected the runs that I wish to plot, I want to be able to select peak ranges and run the PseudoVoigt + BackToBackExponential fitting algorithms on each of the selected peaks, seeing the output.

### Details:
- [ ] Plot should have the option to convert to d-spacing using the DIFA, DIFC, and ZERO values generated from the Calibrate tab

- [x] The user should be able to select the peaks they wish to fit using the plotting widget. 

- [x] The result of the fitting algorithms should be overlaid over the top of the data points. 
- [ ] The results of the fitting should be output and stored for saving later.

### Tests:
- Fit results are saved somehow
- Both fit functions are run on each peak. 
- Two close peaks are still able to be fitted.

### PR DESCRIPTION
**Pull request by RichardWaiteSTFC**
**Description of work.**

Every time a fit is successful the model string will be stored as a `custom setup' in the fit property browser under the name of the workspace fitted. This allows the function to be easily applied to other workspaces.
In addition there is a grouping of workspaces created that stores the fit value and error of parameters of the model for each loaded workspace that reproduces as closely as possible the tables in the hdf output of the legacy software. This aggregates the parameters of repeated functions (typically peaks) such that e.g. if three peaks are fitted then the tables of width, height etc. will have three column corresponding to the each of the three peaks. In order to reduce the number of workspaces the fit results for each parameter are stored in matrix worksapces where X corresponds to the index of the repeated function/peak (i.e. 1-3 for the above example), Y and E correspond to the fit value and error on a given parameter. The spectra correspond to the runs loaded (see the table of the UI fitting tab) - in general different models/functions could be fitted to each run, so when there is a parameter that does not exist for a run, the Y and E fields in the relevant row are filled with NaNs. There is also a table workspace that stores the model string that can be used to load the model into the fitpropertybrowser (Setup > Manage Setup > Load From String).

Deleting or renaming the data workspaces will result in the UI table being refreshed and could result in the rows of the fit tables not matching up to the rows un the logs and UI table. However once a subsequent fit has been successful then the fit table will be refreshed and the rows will match again. I can change this behaviour but I thought it's a bit of an edge case.

**To test:**
(1) Get data to fit by following instructions (1-2) in PR #28393 or downloading the data in PR #29130
(2) Load dat as in PRs above and add some workspaces to the plot (check plot box in the table)
(3) Toggle the fit property browser using the 'Fit' button on the toolbar.
(4) Add some functions (right-click to add peaks etc. as usual) and fit
(5) Check that there is a model stored under the fitted workspace name (Setup > Custom Setup > ws_name)
(6) Check the fit group of workspaces contains a workspace per parameter (naming convention FunctionName_ParameterName)
(7) Check that there is a row per run loaded in the table and that the fit values/errors corresponding to the row of the run fitted are correct
(8) Try fitting another workspace with different functions to the first and check the results as above. If a function parameter doesn't exist in the model of a given run, or the run hasn't been fitted at all, then the Y/E values should be NaN (in the latter instance the model string in the table workspace will be empty).

Partially fixes #27780

Don't merge until PR #29594 is merged

---

#### Reviewer ####

Please comment on the following ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)):

##### Code Review #####

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Review by StephenSmith25 — CHANGES_REQUESTED** [review]
This works well and code additions look good - just a few comments.

File: docs/source/interfaces/Engineering Diffraction.rst [hunk 1]
```diff
@@ -169,14 +169,15 @@ This tab will allow for single peak fitting of focused run files.
...
-containing the average log vlaues) in a nexus format which can also be loaded to provide a guess for the fit to another run.
+The output from the fit is stored in a group of workspaces that contains a matrix workspace of the fit value and error for each parameter in the model. If there is more than one of the same function, the parameters are stored in the same workspace with different x-values. For example, if there were two Gaussian peaks then there would be a workspace for each parameter of the Gaussian (i.e. Height, PeakCentre, Sigma) each of which will have two columns corresponding to each peak. Each workspace has a spectra per run loaded (each row c row in the table of the UI fitting tab). In general different models/functions could be fitted to each run, so when there is a parameter that does not exist for a run (or that run has not yet been fitted), the Y and E fields in the relevant row are filled with NaNs. The group of fit workspaces also contains a table workspace that stores the model string that can be copied into the fitpropertybrowser (Setup > Manage Setup > Load From String).
```
[the remark this answers is not in the document]
Comment (RichardWaiteSTFC) [reply docs/source/interfaces/Engineering Diffraction.rst:180]: Good catch thanks

File: scripts/Engineering/gui/engineering_diffraction/tabs/fitting/data_handling/data_model.py [hunk 2]
```diff
@@ -141,6 +146,71 @@ def update_log_group_name(self):
...
+            axis = TextAxis.create(nruns)
+            for iws, wsname in enumerate(self._loaded_workspaces.keys()):
+                if wsname in self._fit_results:
+                    if param in self._fit_results[wsname]['results']:
+                        fitvals = array(self._fit_results[wsname]['results'][param])
+                        # pad to max length (with nans)
+                        data = vstack((fitvals, full((nfuncs - fitvals.shape[0], 2), nan)))
+                    else:
+                        data = full((nfuncs, 2), nan)
+                else:
```
[the remark this answers is not in the document]
Comment (RichardWaiteSTFC) [reply scripts/Engineering/gui/engineering_diffraction/tabs/fitting/data_handling/data_model.py:192]: Thanks - I'd forgotten the conditions are evaluated in order that makes this possible

**Review by StephenSmith25 — COMMENTED** [review]
Looks good to me, just one thing I noted during testing (which is possibly unrelated) But if you download the file as per the instructions, load the data and click subtract background it brings up the error reporter for me.

**Review by StephenSmith25 — APPROVED** [review]
Having discussed this issue with Richard its outside of the scope. The issue has been outlined in #29823 and will be investigated at another time. Therefore, this is good to go, good work 👍
